Add unit tests for existing reducers and actions, refactor reducers and a few components

Mohammer5 opened the pull request:
This PR contains a few things:

1. I've refactored some of the existing reducers/actions and affected components to be able to reuse state across multiple sections
2. I've added tests for ALL reducers and for all action thunks that load stuff from the api
3. I've adjusted a few components to reflect Hendriks changes (Feedback and App)

HendrikThePendric commented on src/redux/reducers/loading.js:
```diff
@@ -0,0 +1,61 @@
+        default:
+            return state;
+    }
+}
```
Aha, this is where `loading.loading` originates from.... As I said in my previous comment, I think we should distinguish between "fetch-once" and "fetch-repeatably" type of resources here.

Entities like standard-reports and resources, etc. have paging/filtering/deleting, so these actually need a loading state. My personal preference would be to integrate this into the existing reducers for that entity instead of having a generic loading state.

Other entities, like org-units and period-types, are more or less static resources (from the perspective of the reports-app). After they have been fetched nothing happens with them. IMO these don't need a loading state at all. So far all components (I've made) handle the scenario of not loading themselves.

So, if it were up to me I'd like to get rid of this reducer all together.

Mohammer5 replied:
I'll reply here to all the loading state comments you added.

# Why outsourcing the loading state
I outsourced the htmlReport into its own reducer/actions as it's basically reused across the application sections. At first there was a loading state inside this reducer as well, but all section kind of depend on whether something is loading and interaction should be "disabled" when loading to prevent weird behavior (i. e. by showing the loading spinner of the Feedback component). This is also the case for the standard report section, so the `disabled` state for interaction would derive from multiple loading sources which, in my opinion, is an unnecessary overhead if we can simply have one loading that that needs to be set to true/false for the whole application.

# Why having a loading state for the data that's being loaded during app initialization
I think that preventing interaction is also good practice for loading organisation units, data sets in the beginning as well. Why should the user be able to interact with the application while the required data for interaction is being loaded (especially for users with a low-bandwidth internet connection)? At least this was my assumption.

If you agree with this, then checking whether the user can interact or not would derive from "many" loading states of several reducers, which would make the app more complicated than it needs to be.

# Summary of both previous comment section
If preventing the user from interaction with the application is a good idea, and you agree that this is also the case when loading organisation units, data sets, etc., in the beginning, then the loading reducer is a nice way for react to many actions that trigger the loading state to change rather than to spread the loading state across multiple reducers, which results in a more complex and complicated state.

# Why `loading.loading`
The reducer is called loading, that's explains the first `loading`. As requesting an ajax resource can result in an error, I added the property `loadingError` to the state as well. So I wasn't sure how to name the property for the actual `loading` state, which is why there's the name duplication.
We can change this though, I'm not so sure about this anymore as we're storing loading errors in the feedback state as well. Would you agree to change the state from `{ loading: boolean; loadingError: string; }` to `boolean`? Then we'd get rid of the `state.loading.loading`, it'd change to `state.loading`. The loadingError property isn't used anywhere anyways..

HendrikThePendric replied:
I've numbered your headers cause I'm lazy:
**reg. 1 and 4:** I don't necessarily disagree with having a global loading state. It might be a good idea for app/section initialisation. However, I still think we need to distinguish between two scenarios:
a. Entities that are being fetched once (in the app or in a section), and are required for the app (or the app-section) to function. During loading we should disable interaction, so these are related to the central app loading state. OrgUnit, PeriodType, etc. would fall under this.
b. Entities that are fed to list components, such as resource,  shouldn't block the interaction with the app in general, rather, it should just show a spinner on the list component. So in my mind, this would NOT be part of the central loading state.

**reg. 3:** OK, you've won me over: having a central loading state like you have implemented is good. But the fact still remains that the stuff that is being fetched once from `App.js` doesn't need a `LOADING_START` type of action. Initially loading state is false and once the response comes it, it can be set to true.

**reg 4:** It would be nice if loading was just a boolean. Otherwise, we could call the reducer `status` and keep the structure as is. `status.loading` `status.error` that makes sense to me.

**In general**:
I don't think the current implementation of the reducer is logically sound. In theory, the state could get toggled between loading true || false very often, which would then trigger a lot of re-renders. To make it more reliable in the current implementation, we would need some sort of lookup object/Map. The keys would be the action types minus `_START` or `_SUCCESS`, and loading should only be toggled to false of all keys are false.

**An alternative**:
Suppose we ditch the general loading state reducer, and instead we introduce a selector. This selector could be used in the `mapStateToProps` function of the app component, and would read the loading state of separate entities and return a boolean. I think I would actually prefer this approach, because the general loading state is actually a derived state, and this is what selectors are for.....

Mohammer5 replied:
I'll summarize a few things first, then comment on your proposals.

### 1. We need a global loading state for blocking interaction for some requests
These will be org units, period types, pagination (standard report and resource section) and html report requests.

### 2. We have some dropdowns which options will be loaded which shouldn't block app interaction.
These loading states should disable interaction with the dropdown the options are loaded for.

### 3. The loading state for values that are being fetched once and during startup of the app are set to true initially, so we don't need to dispatch loading actions for those.

So regarding the idea to use a selector, I think it's a nice idea. This means:
1. We get rid of the loading reducer
1. All reducers that contain state that must be loaded will have a loading state
1. We have a selector for `blockUi`, which is derived from all the ui blocking loading states
1. We get rid of the loadingError state as that's being handled by the feedback component
1. We display the loading state of dropdown components that don't need to block the whole UI on the dropdowns themselves

I have one remark regarding the last one: 
The DropDown component doesn't have a 'loading' property. So I suggest we disable the dropdown and show 'Options are being loaded' as hint text. Do you agree to this?

HendrikThePendric replied:
Oh, one thing I just realised: we have the `feedback` component which can also show a loader. I suppose this component should be subscribing to the derived loading state, so it can show the loader.... Not the app component as I suggested earlier.
Then it would be something like this in `src/components/Feedback.js` line 50:
```javascript
showLoader: state.feedback.showLoader || blockUI(state),
```

Mohammer5 replied:
yes, that was my plan anyways :)

Mohammer5 replied:
@HendrikThePendric 
I just noticed that it doesn't make much sense to combine the loading and snackbar components.

What I'll do is:
1. Split those components (but keep the feedback state in one reducer)
1. Put the loading component into the App.js (this way we don't have to calculate the showLoader prop for every component)
1. Put the snackbar component into every section (as the standard report and resource section need to add the onActionClick function to the conf)

Mohammer5 commented on src/pages/data-set-report/connectDataSetReport.js:
```diff
@@ -29,9 +30,7 @@ const mapStateToProps = ({
-    loading: dataSetReport.loading,
-    showFeedback: dataSetReport.showFeedback,
-    feedbackConf: dataSetReport.feedbackConf,
+    loading: loading.loading,
```
See my comment at the bottom

Mohammer5 commented on src/pages/reporting-rate-summary/connectReportingRateSummary.js:
```diff
@@ -7,7 +7,7 @@ import {
 
 const mapStateToProps = state => ({
-    loading: state.htmlReport.loading,
+    loading: state.loading.loading,
```
See my comment at the bottom

Mohammer5 commented on src/pages/standard-report/connectStandardReport.js:
```diff
@@ -17,6 +17,7 @@ import {
 
 const mapStateToProps = state => ({
     ...state.standardReport,
+    loading: state.loading.loading,
```
See my comment at the bottom

Mohammer5 commented on src/redux/actions/dataSet/__tests__/asyncThunks.test.js:
```diff
@@ -0,0 +1,81 @@
+            })
+        })
+    })
+})
```
**First my reasons why I went this way:**
1. I generally think when testing units that are a composition of sub units, it's duplicate effort to test whether the sub units return the right stuff and that it should be tested in a separate test for the sub units.
1. I also wrote the tests for the thunks SUPER quick if I'm not testing the actual actions being dispatched (as the thunk doesn't really care about that, I'm more or less interested if the right action creators have been called and their return values have been dispatched).
1. It guarantees that the thunk is working, especially if we omit tests for the action creators themselves as we're expecting them to work (which means that the correct actions will be dispatched is the correct creators are called)
1. We'll have to test for a subset of actions that are being dispatched anyways, because the actions being dispatched from the tested thunk might be thunks themselves and dispatch multiple actions as well, which will 1. increase the amount of work put into creating the test and 2. result in a higher effort when refactoring the actions that are being dispatched. Here's an example: When loading options for a dropdown and it fails, the error action creator is a thunk itself, dispatching also a feedback action. And testing this behavior is definitely not part of the test for the thunk that actually does the ajax request. It should be tested separately in a test for the error action creator.

**Having said that, I see why this might be the wrong approach for the following reasons:**
1. I'm not following the convention on how to test thunks which might seem weird to other devs
1. It will create inconsistency between apps

I still think that not testing the dispatched actions is a better approach as it's quicker to do and requires less test refactoring when working on the actions themselves, but I'd agree to rewrite them for the two mentioned reasons

HendrikThePendric replied:
Thanks for explaining and agreeing to rewrite.
I think it is important to keep testing the actual returned action objects, because essentially the form the contract between the actionCreators and the reducers. Ultimately this is what we want to know: are we dispatching the correct actions to the reducers?

Mohammer5 commented on src/redux/actions/__tests__/organisationUnits.test.js:
```diff
@@ -1,94 +1,192 @@
+
+    describe('organisation units', () => {
+        const mockLoadingOrgUnitsStart = () => ({
+            type: TYPES.ORGANISATION_UNITS_LOADING_START,
```
See my comment regarding the loading state at the bottom

Mohammer5 commented on src/redux/actions/__tests__/organisationUnits.test.js:
```diff
@@ -1,94 +1,192 @@
+    const mockShowErrorFeedback = errorMessage => ({
+        type: feedbackActionTypes.FEEDBACK_SHOW_SNACKBAR,
+        payload: { message: errorMessage, type: feedbackTypes.ERROR },
+    })
```
> By creating these mock-action-creators, we are testing if our mocks do the same as our actual action-creators. It doesn't make sense to me.
This is the same case if we mock the actual action object

The reason why I did this (and I forgot to add this to the test, my bad, so it's not obvious):
We're mocking the response of the ajax request to be `Promise.reject()`.
It should've been the following though:
```js
const error = new Error('Custom error message')
const response = Promise.reject(error);
```

This way we can test whether the error action creator receives the custom error message AND if it falls back to the default one if there's no custom error message or the response is a 404.
This wouldn't be possible if we mock a static action object.

The only difference between mocking a static action object (the expected actions) and dynamic action objects is the payload, which is something I wanted to test explicitly and forgot to do.. My bad, sorry for that, good that you mentioned and found this, so I can add this if you agree with my points.

HendrikThePendric replied:
> This way we can test whether the error action creator receives the custom error message AND if it falls back to the default one if there's no custom error message or the response is a 404.
I'm not sure we need to test that here. I believe we are already testing that in `src/utils/__tests__/humanReadableErrorMessage.test.js`

Mohammer5 commented:
Ok.. Quite a few things have changed now:

1. I adjusted the loading stuff like we discussed
1. I refactored the tests to use the mocked store
1. I took apart the feedback stuff and separated the loading component from the snackbar component while keeping the state in one reducer
1. I removed the display of the loading error from organisation unit and period components as that was redundant, we're displaying the error with the snackbar already.

The only thing that's now missing is the loading/error (disabled + hint text) state for the requests that shouldn't block the UI state.

Mohammer5 commented:
Alright, I made a few more changes, now this one is ready for the next review @HendrikThePendric

HendrikThePendric commented on src/components/DatasetsDropdown.js:
```diff
@@ -16,13 +16,17 @@ export const DatasetsDropdown = props => (
-            hintText={i18n.t('Select Data Set')}
+            hintText={i18n.t(
+                props.loading ? 'Loading options...' : 'Select Data Set'
+            )}
```
I don't think this will work. You need to do something like this:
```javascript
// above the component, after the imports
const hintTextLoading = i18n.t('Loading options...')
const hintText = i18n.t('Select dataset')

// in the component prop
hintText={props.loading ? hintTextLoading : hintText}
```
Because the string is the key, you can't have dynamic strings

HendrikThePendric commented on src/components/feedback/__tests__/__snapshots__/Loader.test.js.snap:
```diff
@@ -0,0 +1,13 @@
+<ContextConsumer>
+  <Component />
+</ContextConsumer>
+`;
```
I've had this happen a few times too. For snapshots you need to make sure to import the un-connected component. Otherwise the output is a bit meaningless

HendrikThePendric commented on src/redux/actions/__tests__/orgUnitGroupSets.test.js:
```diff
@@ -0,0 +1,89 @@
+    const mockLoadingGroupSetsError = errorMessage => ({
+        type: actionTypes.LOADING_GROUP_SETS_ERROR,
+        payload: errorMessage,
+    })
```
In most action tests, you are now directly using with the actual exported actionCreators, which is good. However, here you are still using the mocked functions. Any particular reason for this?
